Extension Defaults

In the SpotBugsExtension constructor you have it setup like this:
```
@Inject
SpotBugsExtension(Project project, ObjectFactory objects) {
	(...)

	project.afterEvaluate( { p ->
		projectName.convention(p.getName());
		release.convention(p.getVersion().toString());
		reportsDir.convention(project.layout.buildDirectory.dir('reports/spotbugs'))
	});

	(...)
}
```

At the time I apply the plugin the project has already been evaluated though and I have to set those values manually. I also don't think that setup supports multi-module builds that use `allprojects{}` or `subprojects{}` to apply a common version.

I think a better setup would be something to the tune of:
```
@Inject
SpotBugsExtension(Project project, ObjectFactory objects) {
	(...)
	
	if (project.state.executed && project.rootProject.state.executed) {
		doTheThing()
	} else if (!project.rootProject.state.executed) {
		project.rootProject.afterEvaluate {
			doTheThing()
		}
	} else {
		project.afterEvaluate {
			doTheThing()
		}
	}
	
	(...)
}

void doTheThing() {
    projectName.convention(project.getName());
    release.convention(project.getVersion().toString());
    reportsDir.convention(project.layout.buildDirectory.dir('reports/spotbugs'))
}
```

@KengoTODA I have the same issue as @mistermage :
```bash
$ ls -a
.		..		.gradle		build.gradle
$ cat build.gradle
plugins {
    id 'java'
    id "jp.skypencil.spotbugs.snom" version "0.4.17"
}
$ gradle spotbugsMain

FAILURE: Build failed with an exception.

* What went wrong:
Task 'spotbugsMain' not found in root project 'test123'.

* Try:
Run gradle tasks to get a list of available tasks. Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 449ms
```
although the task exists if this build.gradle file is used:
```groovy
buildscript {
   repositories {
       maven {
           url "https://plugins.gradle.org/m2/"
       }
   }
   dependencies {
       classpath "gradle.plugin.com.github.spotbugs.snom:spotbugs-gradle-plugin-v2:0.4.17"
   }
}


plugins {
    id 'java'
}
apply plugin: "jp.skypencil.spotbugs.snom"
```

--- a/src/functionalTest/groovy/com/github/spotbugs/snom/MultiProjectFunctionalTest.groovy
+++ b/src/functionalTest/groovy/com/github/spotbugs/snom/MultiProjectFunctionalTest.groovy
@@ -86,6 +86,21 @@ repositories {
         assertEquals(TaskOutcome.SUCCESS, result.task(":sub:spotbugsMain").outcome)
     }
 
+    def "can use project name of sub project"() {
+        when:
+        def result = GradleRunner.create()
+                .withProjectDir(rootDir)
+                .withArguments(':sub:spotbugsMain', '--debug')
+                .withPluginClasspath()
+                .withGradleVersion(version)
+                .forwardOutput()
+                .build()
+
+        then:
+        assertEquals(SUCCESS, result.task(":sub:spotbugsMain").outcome)
+        assertTrue(result.output.contains("-projectName, sub (spotbugsMain)"))
+    }
+
     def "can use toolVersion in subprojects block"() {
         setup:
         buildFile << """

--- a/src/functionalTest/groovy/com/github/spotbugs/snom/MultiProjectFunctionalTest.groovy
+++ b/src/functionalTest/groovy/com/github/spotbugs/snom/MultiProjectFunctionalTest.groovy
@@ -29,6 +29,7 @@ class MultiProjectFunctionalTest extends Specification {
     File rootDir
     File buildFile
     String version = System.getProperty('snom.test.functional.gradle', GradleVersion.current().version)
+    File subBuildFile
 
     @BeforeEach
     def setup() {
@@ -58,7 +59,7 @@ public class Foo {
     }
 }
 """
-        File subBuildFile = new File(subProject, "build.gradle")
+        subBuildFile = new File(subProject, "build.gradle")
         subBuildFile << """
 apply plugin: 'java'
 apply plugin: 'jp.skypencil.spotbugs.snom'
@@ -108,6 +109,28 @@ subprojects {
     spotbugs {
         toolVersion = '4.0.0-RC1'
     }
+}
+"""
+
+        when:
+        def result = GradleRunner.create()
+                .withProjectDir(rootDir)
+                .withArguments(':sub:spotbugsMain', '--debug')
+                .withPluginClasspath()
+                .withGradleVersion(version)
+                .forwardOutput()
+                .build()
+
+        then:
+        assertEquals(SUCCESS, result.task(":sub:spotbugsMain").outcome)
+        assertTrue(result.output.contains("SpotBugs 4.0.0-RC1"))
+    }
+
+    def "can use toolVersion in the subproject"() {
+        setup:
+        subBuildFile << """
+spotbugs {
+    toolVersion = '4.0.0-RC1'
 }
 """
 

--- a/src/functionalTest/groovy/com/github/spotbugs/snom/MultiProjectFunctionalTest.groovy
+++ b/src/functionalTest/groovy/com/github/spotbugs/snom/MultiProjectFunctionalTest.groovy
@@ -21,7 +21,9 @@ import org.gradle.util.GradleVersion
 import org.junit.jupiter.api.BeforeEach
 import spock.lang.Specification
 
+import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
 import static org.junit.jupiter.api.Assertions.assertEquals
+import static org.junit.jupiter.api.Assertions.assertTrue
 
 class MultiProjectFunctionalTest extends Specification {
     File rootDir
@@ -82,5 +84,29 @@ repositories {
         then:
         assertEquals(TaskOutcome.SUCCESS, result.task(":sub:classes").outcome)
         assertEquals(TaskOutcome.SUCCESS, result.task(":sub:spotbugsMain").outcome)
+    }
+
+    def "can use toolVersion in subprojects block"() {
+        setup:
+        buildFile << """
+subprojects {
+    spotbugs {
+        toolVersion = '4.0.0-RC1'
+    }
+}
+"""
+
+        when:
+        def result = GradleRunner.create()
+                .withProjectDir(rootDir)
+                .withArguments(':sub:spotbugsMain', '--debug')
+                .withPluginClasspath()
+                .withGradleVersion(version)
+                .forwardOutput()
+                .build()
+
+        then:
+        assertEquals(SUCCESS, result.task(":sub:spotbugsMain").outcome)
+        assertTrue(result.output.contains("SpotBugs 4.0.0-RC1"))
     }
 }

--- a/src/functionalTest/groovy/com/github/spotbugs/snom/MultiProjectFunctionalTest.groovy
+++ b/src/functionalTest/groovy/com/github/spotbugs/snom/MultiProjectFunctionalTest.groovy
@@ -19,6 +19,7 @@ import org.gradle.testkit.runner.GradleRunner
 import org.gradle.testkit.runner.TaskOutcome
 import org.gradle.util.GradleVersion
 import org.junit.jupiter.api.BeforeEach
+import spock.lang.Ignore
 import spock.lang.Specification
 
 import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
@@ -102,6 +103,7 @@ repositories {
         assertTrue(result.output.contains("-projectName, sub (spotbugsMain)"))
     }
 
+    @Ignore("Gradle does not support this type of configuration. See https://git.io/JvOVT#issuecomment-580239267")
     def "can use toolVersion in subprojects block"() {
         setup:
         buildFile << """

--- a/src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy
+++ b/src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy
@@ -15,6 +15,7 @@ package com.github.spotbugs.snom;
 
 import edu.umd.cs.findbugs.annotations.NonNull
 import edu.umd.cs.findbugs.annotations.Nullable
+import org.gradle.api.Action
 import org.gradle.api.file.DirectoryProperty
 import org.gradle.api.file.RegularFileProperty;
 
@@ -154,11 +155,11 @@ class SpotBugsExtension {
         onlyAnalyze = objects.listProperty(String);
         projectName = objects.property(String);
         release = objects.property(String);
-        project.afterEvaluate( { p ->
-            projectName.convention(p.getName());
-            release.convention(p.getVersion().toString());
+        configureFromProject(project, { p ->
+            projectName.convention(p.getName())
+            release.convention(p.getVersion().toString())
             reportsDir.convention(project.layout.buildDirectory.dir('reports/spotbugs'))
-        });
+        })
         jvmArgs = objects.listProperty(String);
         extraArgs = objects.listProperty(String);
         maxHeapSize = objects.property(String);
@@ -173,5 +174,15 @@ class SpotBugsExtension {
     void setEffort(@Nullable String name) {
         Effort effort = name == null ? null : Effort.valueOf(name.toUpperCase())
         getEffort().set(effort)
+    }
+
+    private void configureFromProject(Project project, Action<Project> action) {
+        if (project.state.executed && project.rootProject.state.executed) {
+            action.execute project
+        } else if (!project.rootProject.state.executed) {
+            project.rootProject.afterEvaluate action
+        } else {
+            project.afterEvaluate action
+        }
     }
 }

